File: backend/models/Blog.js
const mongoose = require('mongoose');

const blogSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Blog title is required'],
    trim: true,
    maxLength: [100, 'Title cannot exceed 100 characters']
  },
  content: {
    type: String,
    required: [true, 'Blog content is required'],
    maxLength: [50000, 'Content cannot exceed 50000 characters']
  },
  excerpt: {
    type: String,
    maxLength: [200, 'Excerpt cannot exceed 200 characters']
  },
  author: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  tags: [{
    type: String,
    trim: true,
    lowercase: true,
    maxLength: [20, 'Tag cannot exceed 20 characters']
  }],
  category: {
    type: String,
    required: [true, 'Blog category is required'],
    trim: true,
    lowercase: true,
    maxLength: [30, 'Category cannot exceed 30 characters']
  },
  featuredImage: {
    type: String,
    default: ''
  },
  status: {
    type: String,
    enum: ['draft', 'published'],
    default: 'published'
  },
  likes: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  bookmarks: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  viewCount: {
    type: Number,
    default: 0
  },
  readTime: {
    type: Number, // in minutes
    default: 1
  },
  isDeleted: {
    type: Boolean,
    default: false
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for like count
blogSchema.virtual('likeCount').get(function() {
  return this.likes.length;
});

// Virtual for bookmark count
blogSchema.virtual('bookmarkCount').get(function() {
  return this.bookmarks.length;
});

// Virtual for comment count
blogSchema.virtual('commentCount', {
  ref: 'Comment',
  localField: '_id',
  foreignField: 'blog',
  count: true
});

// Calculate read time based on content length
blogSchema.pre('save', function(next) {
  if (this.isModified('content')) {
    const wordsPerMinute = 200;
    const words = this.content.split(/\s+/).length;
    this.readTime = Math.max(1, Math.ceil(words / wordsPerMinute));
    
    // Generate excerpt if not provided
    if (!this.excerpt) {
      const plainText = this.content.replace(/<[^>]*>/g, ''); // Remove HTML tags
      this.excerpt = plainText.substring(0, 150) + (plainText.length > 150 ? '...' : '');
    }
  }
  next();
});

// Indexes for better query performance
blogSchema.index({ author: 1, createdAt: -1 });
blogSchema.index({ tags: 1 });
blogSchema.index({ title: 'text', content: 'text', tags: 'text' });
blogSchema.index({ status: 1, isDeleted: 1, createdAt: -1 });

// Helper method to check if user liked the blog
blogSchema.methods.isLikedBy = function(userId) {
  return this.likes.some(like => like.user.toString() === userId.toString());
};

// Helper method to check if user bookmarked the blog
blogSchema.methods.isBookmarkedBy = function(userId) {
  return this.bookmarks.some(bookmark => bookmark.user.toString() === userId.toString());
};

module.exports = mongoose.model('Blog', blogSchema);

File: mobile/lib/models/blog.dart
import 'user.dart';

class Blog {
  final String id;
  final String title;
  final String content;
  final String? excerpt;
  final String? featuredImage;
  final List<String> tags;
  final String category;
  final User author;
  final List<String> likes;
  final int likesCount;
  final int commentsCount;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isBookmarked;
  final bool isLiked;

  Blog({
    required this.id,
    required this.title,
    required this.content,
    this.excerpt,
    this.featuredImage,
    required this.tags,
    required this.category,
    required this.author,
    required this.likes,
    required this.likesCount,
    required this.commentsCount,
    required this.createdAt,
    required this.updatedAt,
    this.isBookmarked = false,
    this.isLiked = false,
  });

  static List<String> _extractUserIds(List<dynamic> items) {
    return items.map((item) {
      if (item is String) {
        return item;
      } else if (item is Map<String, dynamic> && item['user'] != null) {
        return item['user'].toString();
      }
      return '';
    }).where((id) => id.isNotEmpty).toList();
  }

  factory Blog.fromJson(Map<String, dynamic> json) {
    return Blog(
      id: json['_id'] ?? json['id'],
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      excerpt: json['excerpt'],
      featuredImage: json['featuredImage'],
      tags: List<String>.from(json['tags'] ?? []),
      category: json['category'] ?? '',
      author: User.fromJson(json['author'] ?? {}),
      likes: _extractUserIds(json['likes'] ?? []),
      likesCount: json['likeCount'] ?? json['likesCount'] ?? 0,
      commentsCount: json['commentCount'] ?? json['commentsCount'] ?? 0,
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null 
          ? DateTime.parse(json['updatedAt']) 
          : DateTime.now(),
      isBookmarked: json['isBookmarked'] ?? false,
      isLiked: json['isLiked'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'title': title,
      'content': content,
      'excerpt': excerpt,
      'featuredImage': featuredImage,
      'tags': tags,
      'category': category,
      'author': author.toJson(),
      'likes': likes,
      'likesCount': likesCount,
      'commentsCount': commentsCount,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'isBookmarked': isBookmarked,
      'isLiked': isLiked,
    };
  }

  Blog copyWith({
    String? id,
    String? title,
    String? content,
    String? excerpt,
    String? featuredImage,
    List<String>? tags,
    String? category,
    User? author,
    List<String>? likes,
    int? likesCount,
    int? commentsCount,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isBookmarked,
    bool? isLiked,
  }) {
    return Blog(
      id: id ?? this.id,
      title: title ?? this.title,
      content: content ?? this.content,
      excerpt: excerpt ?? this.excerpt,
      featuredImage: featuredImage ?? this.featuredImage,
      tags: tags ?? this.tags,
      category: category ?? this.category,
      author: author ?? this.author,
      likes: likes ?? this.likes,
      likesCount: likesCount ?? this.likesCount,
      commentsCount: commentsCount ?? this.commentsCount,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isBookmarked: isBookmarked ?? this.isBookmarked,
      isLiked: isLiked ?? this.isLiked,
    );
  }
}